From 92b8b400f7bffb2bd605ae57c48888fd785b3fd6 Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hauke.mehrtens@intel.com>
Date: Mon, 8 Jun 2020 17:41:38 +0200
Subject: [PATCH] PONRTSYS-6950: Fix usage of devm_kfree()

When devm_kfree() is called with NULL, it will print a warning because
it can not find the memory. Check if the pointer is NULL before calling
devm_kfree()

Fix a small typo in a comment.
---
 drivers/net/datapath/dpm/gswip31/datapath_gswip.c | 14 +++++++++-----
 1 file changed, 9 insertions(+), 5 deletions(-)

diff --git a/drivers/net/datapath/dpm/gswip31/datapath_gswip.c b/drivers/net/datapath/dpm/gswip31/datapath_gswip.c
index 6d6274d9e5b1..d856ba33b136 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_gswip.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_gswip.c
@@ -1369,7 +1369,7 @@ int dp_meter_del_31(struct net_device *dev,  struct dp_meter_cfg  *meter,
 		pce_rule = devm_kzalloc(&g_dp_dev->dev, sizeof(GSW_PCE_rule_t),
 					GFP_ATOMIC);
 		if (!pce_rule) {
-			pr_err("ctp_cfg alloc failed\n");
+			pr_err("pce_rule alloc failed\n");
 			bret = -1;
 			goto err;
 		}
@@ -1523,10 +1523,14 @@ int dp_meter_del_31(struct net_device *dev,  struct dp_meter_cfg  *meter,
 
 	bret = 0;
 err:
-	devm_kfree(&g_dp_dev->dev, br_cfg);
-	devm_kfree(&g_dp_dev->dev, bp_cfg);
-	devm_kfree(&g_dp_dev->dev, ctp_cfg);
-	devm_kfree(&g_dp_dev->dev, pce_rule);
+	if (br_cfg)
+		devm_kfree(&g_dp_dev->dev, br_cfg);
+	if (bp_cfg)
+		devm_kfree(&g_dp_dev->dev, bp_cfg);
+	if (ctp_cfg)
+		devm_kfree(&g_dp_dev->dev, ctp_cfg);
+	if (pce_rule)
+		devm_kfree(&g_dp_dev->dev, pce_rule);
 	return bret;
 }
 
